code/game/machinery/shieldgen.dm

```diff
@@ -553,3 +553,656 @@
+		return
+	icon_state = "relay_on_[panel_open ?"open":"closed"]"
+
+/obj/machinery/modular_shield/module/relay/RefreshParts()
+	. = ..()
+	range_boost = initial(range_boost)
+	for(var/datum/stock_part/micro_laser/new_laser in component_parts)
+		range_boost += new_laser.tier * 0.25
+
+	if(shield_generator)
+		shield_generator.calculate_boost()
+
+/obj/machinery/modular_shield/module/well
+
+	name = "Modular Shield Well"
+	desc = "A device used to hold more hardlight for the modular shield generator."
+	icon = 'icons/obj/machines/modular_shield_generator.dmi'
+	icon_state = "well_off_closed"
+
+	circuit = /obj/item/circuitboard/machine/modular_shield_well
+
+	///Amount of max strength this machine grants the connected generator
+	var/strength_boost = 0
+
+/obj/machinery/modular_shield/module/well/RefreshParts()
+	. = ..()
+	strength_boost = initial(strength_boost)
+	for(var/datum/stock_part/capacitor/new_capacitor in component_parts)
+		strength_boost += new_capacitor.tier * 10
+
+	if(shield_generator)
+		shield_generator.calculate_boost()
+
+/obj/machinery/modular_shield/module/well/update_icon_state()
+	. = ..()
+	if(!shield_generator || (machine_stat & NOPOWER))
+		icon_state = "well_off_[panel_open ?"open":"closed"]"
+		return
+	icon_state = "well_on_[panel_open ?"open":"closed"]"
+
+
+//The shield itself
+/obj/structure/emergency_shield/modular
+	name = "Modular energy shield"
+	desc = "An energy shield with varying configurations."
+	color = "#00ffff"
+	resistance_flags = INDESTRUCTIBLE //the shield itself is indestructible or atleast should be
+	//our parent
+	var/obj/machinery/modular_shield_generator/shield_generator
+
+
+/obj/structure/emergency_shield/modular/Initialize(mapload)
+	. = ..()
+	AddElement(/datum/element/atmos_sensitive, mapload)
+
+/obj/structure/emergency_shield/modular/should_atmos_process(datum/gas_mixture/air, exposed_temperature)
+	return exposed_temperature > (T0C + 400) //starts taking damage from high temps at the same temperature that nonreinforced glass does
+
+/obj/structure/emergency_shield/modular/atmos_expose(datum/gas_mixture/air, exposed_temperature)
+	if(shield_generator)
+		shield_generator.shield_drain(round(air.return_volume() / 400))//400 integer determines how much damage the shield takes from hot atmos (higher value = less damage)
+		return
+	qdel(src)
+
+
+//How the shield loses strength
+/obj/structure/emergency_shield/modular/take_damage(damage_amount, damage_type = BRUTE, damage_flag = 0, sound_effect = 1, attack_dir)
+	. = ..()
+	if(damage_type == BRUTE || damage_type == BURN)
+		if(shield_generator)
+			shield_generator.shield_drain(damage_amount)//can add or subtract a flat value to buff or nerf crowd damage
+			return
+		qdel(src)
+
+/obj/structure/emergency_shield/modular/emp_act(severity)
+	if(shield_generator)
+		shield_generator.shield_drain(15 / severity) //Light is 2 heavy is 1, note emp is usually a large aoe, tweak the number if not enough damage
+		return
+	qdel(src)
```

Comment:
invert these so that we do the same `isnull(shield_generator)` check and qdel in that tree like i showed in the line 1189 review of this file.